### PR TITLE
More faithful protocol implementation

### PR DESCRIPTION
**Pull request by rgrinberg**
* Remove nuclide related fields
* Reflect optional fields more closely

### CODE REVIEW

**Comment by rgrinberg** [comment]
@andreypopp on second thought, I'm wondering if if we should automatically coerce nulls to empty lists. As far as I understand, this distinction shouldn't matter although it's not specified explicitly anywhere.

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-completion.test.ts [hunk 1]
```diff
@@ -285,17 +285,17 @@ describe("textDocument/completion", () => {
 
   it("completes labels", async () => {
     openDocument(outdent`
-      let f = ListLabels.map 
+      let f = ListLabels.map
```
Comment (andreypopp) [inline ocaml-lsp-server/test/e2e/__tests__/textDocument-completion.test.ts]: Tests now fail because of this

**Comment by rgrinberg** [comment]
> Yeah, shouldn't matter in most cases but there might be some where "absence" is different than "empty". Don't know if there are any in LSP though .

@c-cube what do you think? will this make a difference for you?

I would like it be possible to implement generic LSP servers with this library, so  hopefully this will not matter in practice

**Comment by rgrinberg** [comment]
I'll disable the completion tests for now. Let's write them in a portable manner first, and then we can enable them.

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-completion.test.ts (hunk 1 above)
Comment (rgrinberg) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-completion.test.ts]: Currently, the completion tests aren't in a good shape. Let's fix them first before wasting too many cycles on this.

**Comment by rgrinberg** [comment]
I coerced all nulls to empty lists

**Comment by rgrinberg** [comment]
And I've found an instance where `[]` is not the same as absence. When handling `only` for code actions, `[]` means to present no actions while absence of the field means to present all of them.

**Comment by c-cube** [comment]
> And I've found an instance where `[]` is not the same as absence. `only` for code actions:
> 
>     * `[]` means to present no actions
>     * absence of the field means to present all of them.

I think it's reasonable to have a custom sum type there (basically a list option with `All` or `Only of _ list`) and still adopt the convention that `null` is empty list otherwise. I don't think it'd be problematic for me, after all the exact serialization doesn't matter to users, only the semantics of messages.

**Pull request merged**